### PR TITLE
Enable "codevelop` build workflow

### PR DESCRIPTION
**Pull request by jamiebramwell**
This enables a new build workflow named "codevelop".  You can toggle this with the new CMake option to toggle "codevelop" build, `SERAC_ENABLE_CODEVELOP` with any existing host-config. It uses two git submodules, `axom` and `mfem`, and adds their CMake build systems into ours via `add_subdirectory`.

I did my best to guard against their build systems and make them work as closely to our Spack builds as possible but I can't guarantee anything.

### CODE REVIEW

**Comment by white238** [comment]
i give up for the day... lapack is currently not on a link line even though blas is there....

We need to add more TPL directories to our host-config at the spack stage (Lua, blas, lapack, probably more)

**Comment by white238** [comment]
Progress..

Builds and tests work minus two things:

1)` functional_material_state_test` is failing with this error (plus some repeats):

```
Mesh::GeneratePartitioning(...): edgecut = 11
[INFO] Saving data collection at time: 0 to path: qdata_restart/serac_datacoll 


Verification failed: (f_ind==nf) is false:
 --> Unexpected number of faces.
 ... in function: mfem::H1FaceRestriction::H1FaceRestriction(const mfem::FiniteElementSpace &, const mfem::ElementDofOrdering, const mfem::FaceType)
 ... in file: /usr/WS2/white238/serac/repo/mfem/fem/restriction.cpp:777
```

2) MFEM is adding tests to the test build target even though `MFEM_ENABLE_TESTING` is `OFF`.  They aren't being built, which is good, but they shouldn't be added to `test`.

**Comment by samuelpmishLLNL** [comment]
> 1)` functional_material_state_test` is failing with this error (plus some repeats):
> 
> ```
> Mesh::GeneratePartitioning(...): edgecut = 11
> [INFO] Saving data collection at time: 0 to path: qdata_restart/serac_datacoll 
> 
> 
> Verification failed: (f_ind==nf) is false:
>  --> Unexpected number of faces.
>  ... in function: mfem::H1FaceRestriction::H1FaceRestriction(const mfem::FiniteElementSpace &, const mfem::ElementDofOrdering, const mfem::FaceType)
>  ... in file: /usr/WS2/white238/serac/repo/mfem/fem/restriction.cpp:777
> ```

I can't reproduce this error on my local machine w/ develop, and I can't seem to checkout this branch without incurring a git error:

```
sam@provolone:~/code/serac$ git checkout tpls/essman/mfem_submodule 
fatal: not a git repository: ../.git/modules/axom
fatal: could not reset submodule index
```

but, I previously hit the same error `Verification failed: (f_ind==nf) is false:`, and that was related to the fact that `mfem::ParMesh` neglects to initialize information that `mfem:: H1FaceRestriction` needs until `mfem::ParMesh::ExchangeFaceNbrData()` is called. 

At a glance, it looks like that test _does_ make the call to `mfem::ParMesh::ExchangeFaceNbrData()` when reading the mesh file, though.

-----------------

> MFEM is adding tests to the test build target even though MFEM_ENABLE_TESTING is OFF. They aren't being built, 
> which is good, but they shouldn't be added to test.

https://github.com/mfem/mfem/search?q=add_test

it looks like they are using `add_test` in multiple places without guarding it behind `MFEM_ENABLE_TESTING`. This seems like a good thing we could put in a small PR to mfem (either making those calls `add_executable`, or putting the `add_test` calls in a conditional block).

**Comment by samuelpmishLLNL** [comment]
Josh helped me resolve the issues related to checking out this branch to try and reproduce these errors, but I can't build the project. ~It looks like we are propagating our pedantic warnings-as-errors flags down to MFEM, and its code generates a lot of errors as a result (tested with GCC 10.2 and clang 10.0 on my local machine).~

edit: I'm still seeing a lot of errors, but I'm not sure about their origin

**Comment by jamiebramwell** [comment]
> 
> 
> Progress..
> 
> Builds and tests work minus two things:
> 
> 1)` functional_material_state_test` is failing with this error (plus some repeats):
> 
> ```
> Mesh::GeneratePartitioning(...): edgecut = 11
> [INFO] Saving data collection at time: 0 to path: qdata_restart/serac_datacoll 
> 
> 
> Verification failed: (f_ind==nf) is false:
>  --> Unexpected number of faces.
>  ... in function: mfem::H1FaceRestriction::H1FaceRestriction(const mfem::FiniteElementSpace &, const mfem::ElementDofOrdering, const mfem::FaceType)
>  ... in file: /usr/WS2/white238/serac/repo/mfem/fem/restriction.cpp:777
> ```
> 
>     1. MFEM is adding tests to the test build target even though `MFEM_ENABLE_TESTING` is `OFF`.  They aren't being built, which is good, but they shouldn't be added to `test`.

Yes, this normally pops up when we don't call `ExchangeFaceNbrData()` on the mesh.

**Comment by white238** [comment]
I swapped over the git submodule to my fork of MFEM until my PR gets merged. Going forward we will want to avoid forks and just run off MFEM branches.  Mistakes were made.

Here is how to swap a submodule to a different remote:

```
git submodule deinit -f mfem
rm -rf .git/modules/mfem/
git submodule update --init --recursive
```

You only have to do this if you have previously pulled this branch.

**Comment by white238** [comment]
I swapped this back to the mfem repo.  So if you didn't pull between when i went to the fork and now you shouldnt have a problem.

**Comment by white238** [comment]
@samuelpmishLLNL can you try this branch out?  You should be able to do this with a normal host-config and turning on the CMake option `SERAC_ENABLE_CODEVELOP`.

**Comment by samuelpmishLLNL** [comment]
I checked out this branch, did

```
git submodule init
git submodule update
```

configure with

```
/usr/bin/cmake --no-warn-unused-cli -C /home/sam/code/serac/.vscode/serac_libs_debug.cmake -DENABLE_DOCS=ON -DENABLE_CLANGFORMAT=ON -DCLANGFORMAT_EXECUTABLE=/usr/bin/clang-format -DSERAC_ENABLE_CODE_CHECKS=ON -DSPHINX_EXECUTABLE=/usr/bin/sphinx-build -DDOXYGEN_EXECUTABLE=/usr/bin/doxygen -DSERAC_ENABLE_CODEVELOP=ON -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_C_COMPILER:FILEPATH=/bin/x86_64-linux-gnu-gcc-10 -DCMAKE_CXX_COMPILER:FILEPATH=/bin/x86_64-linux-gnu-g++-10 -H/home/sam/code/serac -B/home/sam/code/serac/build -G Ninja
```

but cmake prints

```
[cmake] -- Using installed MFEM
[cmake] -- Looking for MFEM using MFEM_DIR = /home/sam/code/serac_libs_debug/gcc-9.3.0/mfem-4.3.0
[cmake] -- Could NOT find MFEM (missing: MFEM_DIR)
[cmake] -- Using MFEM's GNU Make config file: /home/sam/code/serac_libs_debug/gcc-9.3.0/mfem-4.3.0/share/mfem/config.mk
[cmake] -- Found MFEM: /home/sam/code/serac_libs_debug/gcc-9.3.0/mfem-4.3.0/lib/libmfem.a;-Wl, ...
[cmake] -- MFEM Includes: /home/sam/code/serac_libs_debug/gcc-9.3.0/mfem-4.3.0/include ...
[cmake] -- MFEM Libraries: /home/sam/code/serac_libs_debug/gcc-9.3.0/mfem-4.3.0/lib/libmfem.a; ...
[cmake] -- Using installed Axom
[cmake] -- Axom support is ON
```

which makes it appear that cmake is not using the submodules for mfem/axom. But when actually build serac, it seems almost everything works, execept that I get errors from some mfem headers:

```
[build] /home/sam/code/serac/mfem/linalg/dtensor.hpp:136:21: error: conversion from ‘long unsigned int’ to ‘int’ may change value [-Werror=conversion]
[build]   136 | inline DeviceTensor<sizeof...(Dims),T> Reshape(T *ptr, Dims... dims)
[build]       |                     ^~~~~~~~~~~~~~~
[build] /home/sam/code/serac/mfem/linalg/dtensor.hpp: In function ‘mfem::DeviceTensor<sizeof... (Dims), T> mfem::Reshape(T*, Dims ...)’:
[build] /home/sam/code/serac/mfem/linalg/dtensor.hpp:138:24: error: conversion from ‘long unsigned int’ to ‘int’ may change value [-Werror=conversion]
[build]   138 |    return DeviceTensor<sizeof...(Dims),T>(ptr, dims...);
[build]       |                        ^~~~~~~~~~~~~~~
```

note: the paths for the header files correspond to the mfem submodule.

So it looks like some things are working, but our `-Werror` flag is being applied to sourcefiles outside of our control, and that prevents me from successfully building serac.

**Comment by samuelpmishLLNL** [comment]
update: after discussing these issues with Chris and a full TPL rebuild, things appear to be working

**Review by samuelpmishLLNL — APPROVED** [review]

**Comment by samuelpmishLLNL** [comment]
I am seeing a lot of build targets from mfem (I can't tell if they're being built in by the `all` target yet, due to some unrelated compilation errors on my branch):
Screen Shot 2021-11-17 at 1 53 20 PM

**Comment by white238** [comment]
> 
> 
> I am seeing a lot of build targets from mfem (I can't tell if they're being built in by the `all` target yet, due to some unrelated compilation errors on my branch): 

mfem creates a lot of targets then doesn't add them to `all`.  I have some small fixes for that but it goes against their build target outlook on life.

File: scripts/spack/packages/serac/package.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ class Serac(CachedCMakePackage, CudaPackage):
     depends_on("netcdf-c@4.7.4~shared", when="+netcdf")
 
     # Needs to be first due to a bug with the Spack concretizer
-    depends_on("hdf5+hl@1.8.21~shared")
+    depends_on("hdf5@1.8.21+hl~mpi~shared")
```
Comment (joshessman-llnl) [inline scripts/spack/packages/serac/package.py:112]: comment here?

**Review by joshessman-llnl — APPROVED** [review]

**Review by jamiebramwell — COMMENTED** [review]
Works on my machine. Thanks for figuring this out!! It will make my life a lot easier.

**Comment by jamiebramwell** [comment]
I approve, but I can't approve my own PR. 😆

**Pull request merged**